libgnucash/engine/Account.cpp

```diff
@@ -3375,15 +3396,37 @@ GetBalanceAsOfDate (Account *acc, time64 date, gboolean ignclosing)
     xaccAccountRecomputeBalance (acc); /* just in case, normally a noop */
 
     priv = GET_PRIVATE(acc);
-    if (ignclosing)
-        balance = priv->noclosing_balance;
-    else
+    switch (baltype) {
+    case BalanceType::FULL:
         balance = priv->balance;
+        lp = priv->splits;
+        break;
+    case BalanceType::NO_CLOSING:
+        balance = priv->noclosing_balance;
+        lp = priv->splits;
+        break;
+    case BalanceType::RECONCILED:
+        GList *YREC_splits = NULL;
+        balance = priv->reconciled_balance;
+        for (lp = priv->splits; lp; lp = lp->next)
+        {
+            if (xaccSplitGetReconcile ((Split*)lp->data) == YREC)
+                YREC_splits = g_list_prepend (YREC_splits, lp->data);
+        }
+        YREC_splits = g_list_reverse (YREC_splits);
+        lp = g_list_sort (YREC_splits, (GCompareFunc)compare_reconciled_date);
```

Comment:
I just thought of a serious speedup for this bit. No need to sort the splits or even create a new list:
```C++
             balance = gnc_numeric_zero();
             for (GList *node = priv->splits; node ; node = g_list_next(node))
                  if (xaccSplitGetReconcile((Split*)node->data) == YREC && 
                       xaccSplitGetDateReconciledSec((Split*)node->data) <= date))
                       balance = gnc_numeric_add(balance, xaccSplitGetAmount((Split*)node-data));
              return balance;
```
I think it would make more sense to simply have a separate `get_reconciled_balance_as_of_date(Account*, time64 date)` and leave `GetBalanceAsOfDate(Account*, gboolean)` alone since the logic is substantially different.

Reply:
Done. Reverted Account.cpp commits and recreated in last commit.